File: backend/core/routers/data_governance.py

# [AI GENERATED]
# Model: GitHub Copilot (GPT-4.1)
# Logic: Scaffold a FastAPI router for Data Governance & Stewardship, providing endpoints to manage data owners, stewards, and governance policies.
# Why: Roadmap item 25 requires clear data ownership and stewardship roles for compliance and accountability.
# Root Cause: No API existed for managing or querying data governance roles and policies.
# Context: This router provides endpoints for registering owners/stewards, listing assignments, and managing governance policies. Future: integrate with persistent storage, add policy enforcement, and support audit trails. For advanced governance logic, consider using a more advanced model (Claude Opus).

from fastapi import APIRouter, status, HTTPException
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

router = APIRouter()

# In-memory store for demo purposes
GOVERNANCE_ASSIGNMENTS = []
GOVERNANCE_POLICIES = []

class GovernanceAssignment(BaseModel):
    id: int
    dataset: str
    owner: str
    stewards: List[str]
    assigned_at: datetime

class GovernancePolicy(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    rules: dict
    created_at: datetime

@router.post("/data-governance/assignments", status_code=status.HTTP_201_CREATED)
def register_assignment(assignment: GovernanceAssignment):
    if any(a.id == assignment.id for a in GOVERNANCE_ASSIGNMENTS):
        raise HTTPException(status_code=400, detail="Assignment ID already exists")
    GOVERNANCE_ASSIGNMENTS.append(assignment)
    return {"message": "Assignment registered", "assignment_id": assignment.id}

@router.get("/data-governance/assignments", response_model=List[GovernanceAssignment])
def list_assignments(dataset: Optional[str] = None):
    if dataset:
        return [a for a in GOVERNANCE_ASSIGNMENTS if a.dataset == dataset]
    return GOVERNANCE_ASSIGNMENTS

@router.post("/data-governance/policies", status_code=status.HTTP_201_CREATED)
def register_policy(policy: GovernancePolicy):
    if any(p.id == policy.id for p in GOVERNANCE_POLICIES):
        raise HTTPException(status_code=400, detail="Policy ID already exists")
    GOVERNANCE_POLICIES.append(policy)
    return {"message": "Policy registered", "policy_id": policy.id}

@router.get("/data-governance/policies", response_model=List[GovernancePolicy])
def list_policies():
    return GOVERNANCE_POLICIES

File: services/frontend/src/services/dataGovernance.js
import axios from 'axios';

const API_BASE = '/api';

export const registerAssignment = async (assignment) => {
    const res = await axios.post(`${API_BASE}/data-governance/assignments`, assignment);
    return res.data;
};

export const listAssignments = async (dataset) => {
    const url = dataset ? `${API_BASE}/data-governance/assignments?dataset=${encodeURIComponent(dataset)}` : `${API_BASE}/data-governance/assignments`;
    const res = await axios.get(url);
    return res.data;
};

export const registerPolicy = async (policy) => {
    const res = await axios.post(`${API_BASE}/data-governance/policies`, policy);
    return res.data;
};

export const listPolicies = async () => {
    const res = await axios.get(`${API_BASE}/data-governance/policies`);
    return res.data;
};
